### PR TITLE
Improve history table with previous_groups_id field and counter field

### PR DESCRIPTION
**Pull request by Badoche1337**
Hello,

I want to add some informations in escalation history table :

Previous group assigned (Group A)
Counter of escalations between Group A (previous) and Group B (new assigned group)
New search option : total number of escalations in one ticket
New search option : total number of escalations in one ticket between 2 distincts groups

### CODE REVIEW

File: hook.php [hunk 1]
```diff
@@ -252,6 +252,51 @@ function plugin_escalade_install() {
...
+         $histories[$tickets_id][] = $data;
+      }
+
+      foreach ($histories as $tickets_id => $h) {
+         echo 'Ticket #' . $tickets_id . '<br/>';
+         $counters = [];
+
+         foreach ($h as $k => $details) {
+            if (isset($h[$k+1])) {
+               $first  = $h[$k+1]['groups_id'] < $details['groups_id'] ? $h[$k+1]['groups_id'] : $details['groups_id'];
```
Comment (cedric-anne) [inline hook.php]: I do not understand why there is such a numeric comparison on `groups_id` used to determines `$first` and `$second` values.
If `counter` is the number of escalade for one group, it should only be related to update from this group to another, right ?

File: hook.php [hunk 2]
```diff
@@ -386,6 +431,30 @@ function plugin_escalade_getAddSearchOptions($itemtype) {
...
+         ];
+
+         $sopt[] = [
+            'id'                 => '1992',
+            'table'              => 'glpi_plugin_escalade_histories',
+            'field'              => 'counter',
+            'name'               => __("Number of escalades for one group", "escalade"),
+            'joinparams'         => [
+               'jointype'           => 'child'
+            ]
```
Comment (cedric-anne) [inline hook.php]: `'datatype' => 'integer'` ?

File: inc/history.class.php [hunk 3]
```diff
@@ -24,6 +24,11 @@ static function getlastLineForTicket($tickets_id) {
       }
    }
 
+   static function getLastHistoryForTicketAndGroup($tickets_id, $groups_id, $previous_groups_id) {
+      $history = new self();
+      return $history->find("tickets_id = $tickets_id AND ((groups_id = $groups_id AND previous_groups_id = $previous_groups_id) OR (groups_id = $previous_groups_id AND previous_groups_id = $groups_id))", "date_mod DESC", 1);
```
Comment (cedric-anne) [inline inc/history.class.php]: It could be better to use new querying syntax, instead of raw SQL.

File: inc/ticket.class.php [hunk 4]
```diff
@@ -210,9 +210,30 @@ static function addHistoryOnAddGroup(CommonDBTM $item) {
 
       //add line in history table
       $history = new PluginEscaladeHistory();
+
+      $group_ticket      = new Group_Ticket();
+      $last_group_tickets = $group_ticket->find("tickets_id = $tickets_id AND type = 2", "id DESC", 1);
```
Comment (cedric-anne) [inline inc/ticket.class.php]: Same remark.

File: hook.php [hunk 5]
```diff
@@ -386,6 +431,30 @@ function plugin_escalade_getAddSearchOptions($itemtype) {
...
+         ];
+
+         $sopt[] = [
+            'id'                 => '1992',
+            'table'              => 'glpi_plugin_escalade_histories',
+            'field'              => 'counter',
+            'name'               => __("Number of escalades for one group", "escalade"),
```
Comment (cedric-anne) [inline hook.php]: `escalades` -> `escalations` ?

File: inc/ticket.class.php [hunk 6]
```diff
@@ -210,9 +210,30 @@ static function addHistoryOnAddGroup(CommonDBTM $item) {
...
+
+      $previous_groups_id = 0;
+      $counter            = 0;
+
+      if (count($last_group_tickets) > 0) {
+         $last_group_ticket  = array_shift($last_group_tickets);
+         $previous_groups_id = $last_group_ticket['groups_id'];
+
```
Comment (cedric-anne) [inline inc/ticket.class.php]: I do not know how the plusing works, but maybe you should skpi next lines if `$last_group_ticket['groups_id'] == $groups_id` as it does not looks like an escalation.

File: hook.php (hunk 1 above)
Comment (Badoche1337) [reply hook.php]: The counter field is not number of escalations for the current group (this could be done in simple query), but the number of escalations betwteen current group (A) and the previous one (B), the numeric comparaison is just one way to ensure the key (in array) used to store current count of escalations betwteen the 2 current groups is the same, whatever which group has escalated to other in the current row.

Comment (cedric-anne) [reply hook.php]: OK, so it is only the counter for exact same escalation path if I understand right. Maybe the name "counter" is not enough precise in this case.

Comment (Badoche1337) [reply hook.php]: yes it's the counter of total escalations for A > B or B > A

File: locales/fr_FR.po [hunk 7]
```diff
@@ -35,11 +35,19 @@ msgstr "paramètres manquants"
 msgid "full assignation history"
 msgstr "historique d'assignation complet"
 
-#: hook.php:376
+#: hook.php:422
```
Comment (trasher) [inline locales/fr_FR.po]: PO files should not be part of the PR.

**Comment by Badoche1337** [comment]
Can we merge ? x)
